st2actions/st2actions/scheduler/entrypoint.py

```diff
@@ -61,28 +78,27 @@ def process(self, request):
             LOG.exception('Failed to find liveaction %s in the database.', request.id)
             raise
 
-        # Apply policies defined for the action.
-        liveaction_db = policy_service.apply_pre_run_policies(liveaction_db)
+        query = {
+            "liveaction": str(liveaction_db.id),
+        }
 
-        # Exit if the status of the request is no longer runnable.
-        # The status could have be changed by one of the policies.
-        if liveaction_db.status not in [action_constants.LIVEACTION_STATUS_REQUESTED,
-                                        action_constants.LIVEACTION_STATUS_SCHEDULED]:
-            LOG.info('%s is ignoring %s (id=%s) with "%s" status after policies are applied.',
-                     self.__class__.__name__, type(request), request.id, liveaction_db.status)
+        if ExecutionQueue.query(**query):
             return
 
-        # Update liveaction status to "scheduled".
-        if liveaction_db.status == action_constants.LIVEACTION_STATUS_REQUESTED:
+        if liveaction_db.delay > 0:
             liveaction_db = action_service.update_status(
-                liveaction_db, action_constants.LIVEACTION_STATUS_SCHEDULED, publish=False)
+                liveaction_db,
+                action_constants.LIVEACTION_STATUS_DELAYED,
+                publish=False
+            )
 
-        # Publish the "scheduled" status here manually. Otherwise, there could be a
-        # race condition with the update of the action_execution_db if the execution
-        # of the liveaction completes first.
-        LiveAction.publish_status(liveaction_db)
+        execution_request = _create_execution_request_from_liveaction(
+            liveaction_db,
+            delay=liveaction_db.delay
+        )
+        ExecutionQueue.add_or_update(execution_request, publish=False)
```

Comment:
I totally miss this because I thought this was indented under `if liveaction_db.delay > 0`.  So does this means that every single action execution will be queued into MongoDB?  I thought we agree only to push it into the queue if there's a delay. If there's no delay, then just set it to `scheduled` and pass it to the policy engine.

Reply:
No, that's was never the case. Nor do I really think it should be. Given where we want the scheduler to go all executions should go into queue to be scheduled.

Reply:
So this was my original thought on the subject but then @lakshmi-kannan disagrees.  https://github.com/StackStorm/discussions/issues/316

Then you, @Kami, and I talked over this in a design discussion. And I thought we agree to push only executions with special scheduling conditions into the table.  I thought this was an appropriate compromise.  Maybe we didn't clarify.  https://stackstorm.slack.com/archives/C81BGM60H/p1539793063000200

Anyway, I'm cool with pushing all the executions to a table because that was my original thought. But I prefer this gets a clear from @Kami and @lakshmi-kannan.  Also, if this is the case, we need to process the scheduling in bulk otherwise, the performance and latency will hit us.

Reply:
Current implementation schedules in bulk any available executions ready for scheduling on every round of the loop.